pkg/logger/logger_cluster.go

```diff
@@ -0,0 +1,156 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package logger
+
+import (
+	"sync"
+	"time"
+)
+
+//--
+
+//Cluster defines a service that is capable of clustering loggers into on syncronized output
+type Cluster interface {
+	//GetLoggerPipeline returns the cluster instance the Cluster send the coupled ChannelMessages to
+	GetLoggerPipeline() Pipeline
+
+	//GetChannelProvider returns the channel provider this Cluster uses
+	GetChannelProvider() ChannelProvider
+
+	//StartListening enables the Cluster instance to listen to the channel found in Cluster#getChannelProvider
+	StartListening()
+
+	//Flushes the cluster content to the pipeline
+	Flush()
+
+	//GetWaitGroup returns the waitgroup instance the cluster uses to identify if it is done running
+	GetWaitGroup() *sync.WaitGroup
+}
+
+//simpleCluster is a default implementation of the Cluster interface
+type simpleCluster struct {
+	pipe            Pipeline
+	channel         ChannelProvider
+	clusterDuration time.Duration
+	waitGroup       *sync.WaitGroup
+	messageCache    []ChannelMessage
+	nextCacheFlush  time.Time
+}
+
+//GetLoggerPipeline returns the upstream pipeline the Cluster sends all its data to
+func (s *simpleCluster) GetLoggerPipeline() Pipeline {
+	return s.pipe
+}
+
+//GetChannelProvider returns the channel provider the Cluster wi
+func (s *simpleCluster) GetChannelProvider() ChannelProvider {
+	return s.channel
+}
+
+//StartListening enables the Cluster instance to listen to the channel found in Cluster#getChannelProvider
+//This will block all code executed, so use it in a go routine
+func (s *simpleCluster) StartListening() {
+	s.GetWaitGroup().Add(1)
+
+	for incomingMessage := range s.GetChannelProvider().GetChannel() {
+		if s.timedFlushNeeded() {
+			if s.isLoggerMessageCached(incomingMessage.GetLogger().GetID()) { //The message is not flushed this release
+				s.Flush()
+				s.resetFlushTimer()
+
+				s.appendMessage(incomingMessage)
```

Comment:
Matthias would point out that for readability reasons, the empty line should be after `s.appendMessage(incomingMessage)` so that the grouping is more aligned with the if case.